jax/collect_profile.py

```diff
@@ -0,0 +1,90 @@
+# Copyright 2022 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import argparse
+import gzip
+import os
+import pathlib
+import tempfile
+
+import jax
+try:
+  from tensorflow.python.profiler import profiler_v2 as profiler
+  from tensorflow.python.profiler import profiler_client
+except ImportError:
+  raise ImportError("This script requires `tensorflow` to be installed.")
+try:
+  from tensorboard_plugin_profile.convert import raw_to_tool_data as convert
+except ImportError:
+  raise ImportError(
+      "This script requires `tensorboard_plugin_profile` to be installed.")
+
+
+_DESCRIPTION = """
+To profile running JAX programs, you first need to start the profiler server
+in the program of interest. You can do this via
+`jax.profiler.start_server(<port>)`. Once the program is running and the
+profiler server has started, you can run `collect_profile` to trace the execution
+for a provided duration. The trace file will be dumped into a directory
+(determined by `--log_dir`) and by default, a Perfetto UI link will be generated
+to view the resulting trace.
+"""
+parser = argparse.ArgumentParser(description=_DESCRIPTION)
+parser.add_argument("--log_dir", default=None,
+                    help=("Directory to store log files. "
+                          "Uses a temporary directory if none provided."),
+                    type=str)
+parser.add_argument("port", help="Port to collect trace", type=int)
+parser.add_argument("duration_in_ms",
+                    help="Duration to collect trace in milliseconds", type=int)
+parser.add_argument("--no_perfetto_link",
+                    help="Disable creating a perfetto link", action="store_true")
+parser.add_argument("--host", default="127.0.0.1",
+                    help="Host to collect trace. Defaults to 127.0.0.1",
+                    type=str)
+parser.add_argument("--host_tracer_level", default=2,
+                    help="Profiler host tracer level", type=int)
+parser.add_argument("--device_tracer_level", default=1,
+                    help="Profiler device tracer level", type=int)
+parser.add_argument("--python_tracer_level", default=0,
+                    help="Profiler Python tracer level", type=int)
+
+def main(args):
```

Comment:
To test this, could you pull this out to a function that doesn't require flags, and write a unit test that starts a long-running computation and manually calls this? It might be a big heavy duty.

Reply:
Done, though the test may not actually be run w/o the proper deps: cc @yashk2810